File: server/internal/model/blog.go
package model

import (
	"time"
	"valley-server/internal/utils"

	"gorm.io/gorm"
)

// Post 博客/图文内容
// 注意：CoverStorageKey 用于封面文件替换/删除时做对象存储清理。
type Post struct {
	ID              Int64String    `gorm:"primaryKey;autoIncrement:false" json:"id"`
	Title           string         `gorm:"size:200;not null" json:"title"`
	Slug            string         `gorm:"size:200;uniqueIndex;not null" json:"slug"`
	PostType        string         `gorm:"size:20;default:'blog';index" json:"postType"`
	Visibility      string         `gorm:"size:20;default:'private';index" json:"visibility"`
	TemplateKey     string         `gorm:"size:64" json:"templateKey,omitempty"`
	TemplateData    string         `gorm:"type:text" json:"templateData,omitempty"`
	ImageTextData   string         `gorm:"type:json" json:"imageTextData,omitempty"`
	Content         string         `gorm:"type:text;not null" json:"content"`
	HTMLContent     string         `gorm:"type:text" json:"htmlContent,omitempty"`
	DraftData       string         `gorm:"type:text" json:"draftData,omitempty"`
	Excerpt         string         `gorm:"size:500" json:"excerpt"`
	Cover           string         `gorm:"size:500" json:"cover,omitempty"`
	CoverStorageKey string         `gorm:"size:500" json:"coverStorageKey,omitempty"`
	AuthorID        Int64String    `gorm:"index" json:"authorId"`
	GroupID         Int64String    `gorm:"index" json:"groupId"`
	CategoryID      Int64String    `gorm:"index" json:"categoryId"`
	Status          string         `gorm:"size:20;default:'draft';index" json:"status"`
	ViewCount       int            `gorm:"default:0" json:"viewCount"`
	LikeCount       int            `gorm:"default:0" json:"likeCount"`
	IsTop           bool           `gorm:"default:false" json:"isTop"`
	SortOrder       int            `gorm:"default:0;index" json:"sortOrder"`
	GroupSortOrder  int            `gorm:"default:0;index" json:"groupSortOrder"`
	PublishedAt     *time.Time     `json:"publishedAt,omitempty"`
	DraftUpdatedAt  *time.Time     `json:"draftUpdatedAt,omitempty"`
	CreatedAt       time.Time      `json:"createdAt"`
	UpdatedAt       time.Time      `json:"updatedAt"`
	DeletedAt       gorm.DeletedAt `gorm:"index" json:"-"`

	Author   *User         `gorm:"foreignKey:AuthorID" json:"author,omitempty"`
	Group    *PostGroup    `gorm:"foreignKey:GroupID" json:"group,omitempty"`
	Category *PostCategory `gorm:"foreignKey:CategoryID" json:"category,omitempty"`
	Tags     []PostTag     `gorm:"many2many:post_tag_relations;" json:"tags,omitempty"`
}

func (p *Post) BeforeCreate(tx *gorm.DB) error {
	if p.ID == 0 {
		p.ID = Int64String(utils.GenerateID())
	}
	return nil
}

type PostCategory struct {
	ID          Int64String    `gorm:"primaryKey;autoIncrement:false" json:"id"`
	Name        string         `gorm:"size:50;uniqueIndex;not null" json:"name"`
	Slug        string         `gorm:"size:50;uniqueIndex;not null" json:"slug"`
	Description string         `gorm:"size:255" json:"description"`
	SortOrder   int            `gorm:"default:0" json:"sortOrder"`
	PostCount   int            `gorm:"default:0" json:"postCount"`
	CreatedAt   time.Time      `json:"createdAt"`
	UpdatedAt   time.Time      `json:"updatedAt"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
}

func (c *PostCategory) BeforeCreate(tx *gorm.DB) error {
	if c.ID == 0 {
		c.ID = Int64String(utils.GenerateID())
	}
	return nil
}

type PostGroup struct {
	ID          Int64String    `gorm:"primaryKey;autoIncrement:false" json:"id"`
	Name        string         `gorm:"size:80;not null" json:"name"`
	Slug        string         `gorm:"size:100;not null;uniqueIndex" json:"slug"`
	GroupType   string         `gorm:"size:20;default:'blog';index" json:"groupType"`
	Description string         `gorm:"size:255" json:"description"`
	AuthorID    Int64String    `gorm:"index;not null" json:"authorId"`
	ParentID    *Int64String   `gorm:"index" json:"parentId,omitempty"`
	SortOrder   int            `gorm:"default:0" json:"sortOrder"`
	PostCount   int            `gorm:"default:0" json:"postCount"`
	CreatedAt   time.Time      `json:"createdAt"`
	UpdatedAt   time.Time      `json:"updatedAt"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`

	Author *User      `gorm:"foreignKey:AuthorID" json:"author,omitempty"`
	Parent *PostGroup `gorm:"foreignKey:ParentID" json:"parent,omitempty"`
}

func (g *PostGroup) BeforeCreate(tx *gorm.DB) error {
	if g.ID == 0 {
		g.ID = Int64String(utils.GenerateID())
	}
	return nil
}

type PostTag struct {
	ID        Int64String    `gorm:"primaryKey;autoIncrement:false" json:"id"`
	Name      string         `gorm:"size:30;uniqueIndex;not null" json:"name"`
	Slug      string         `gorm:"size:30;uniqueIndex;not null" json:"slug"`
	PostCount int            `gorm:"default:0" json:"postCount"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

func (t *PostTag) BeforeCreate(tx *gorm.DB) error {
	if t.ID == 0 {
		t.ID = Int64String(utils.GenerateID())
	}
	return nil
}

type PostTagRelation struct {
	PostID Int64String `gorm:"primaryKey;index" json:"postId"`
	TagID  Int64String `gorm:"primaryKey;index" json:"tagId"`
}

func (PostTagRelation) TableName() string {
	return "post_tag_relations"
}

type PostComment struct {
	ID        Int64String    `gorm:"primaryKey;autoIncrement:false" json:"id"`
	PostID    Int64String    `gorm:"index;not null" json:"postId"`
	UserID    Int64String    `gorm:"index;not null" json:"userId"`
	Content   string         `gorm:"size:2000;not null" json:"content"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	Post *Post `gorm:"foreignKey:PostID" json:"post,omitempty"`
	User *User `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

func (c *PostComment) BeforeCreate(tx *gorm.DB) error {
	if c.ID == 0 {
		c.ID = Int64String(utils.GenerateID())
	}
	return nil
}

File: apps/web/src/api/blog.ts
import request, { type RequestConfig } from '@/utils/request';

export type PostType = 'blog' | 'image_text';
export type Visibility = 'private' | 'shared' | 'public';
export type GroupType = PostType;

export interface ImageTextPage {
  imageUrl?: string;
  imageKey?: string;
  text?: string;
  highlightText?: string;
  highlightStart?: number;
  highlightEnd?: number;
  fontSize?: number;
  highlightFontSize?: number;
  textStyles?: Array<{
    start?: number;
    end?: number;
    fontSize?: number;
  }>;
}

export interface ImageTextData {
  templateKey?: string;
  partner?: string;
  stickerEmoji?: string;
  images?: string[];
  pages?: ImageTextPage[];
  generatedAt?: string;
}

export interface PostCategory {
  id: string;
  name: string;
  slug: string;
}

export interface PostGroup {
  id: string;
  name: string;
  slug?: string;
  groupType?: GroupType;
  description?: string;
  authorId: string;
  parentId?: string;
  sortOrder?: number;
  postCount?: number;
}

export interface PostTag {
  id: string;
  name: string;
  slug: string;
}

export interface Post {
  id: string;
  title: string;
  slug?: string;
  postType: PostType;
  visibility?: Visibility;
  templateKey?: string;
  templateData?: string;
  imageTextData?: string;
  excerpt: string;
  cover?: string;
  coverStorageKey?: string;
  groupId: string;
  group?: PostGroup;
  categoryId: string;
  category?: PostCategory;
  tags?: PostTag[];
  status?: 'draft' | 'published' | 'archived';
  viewCount: number;
  likeCount: number;
  isTop: boolean;
  publishedAt?: string;
  createdAt: string;
  author?: {
    id: string;
    nickname: string;
    avatar: string;
  };
}

export interface PostDetail extends Post {
  content: string;
  htmlContent: string;
}

export interface PostComment {
  id: string;
  postId: string;
  userId: string;
  content: string;
  createdAt: string;
  updatedAt: string;
  author?: {
    id: string;
    nickname: string;
    avatar: string;
  };
}

export interface PostCommentListData {
  list: PostComment[];
  total: number;
  postType: PostType;
}

export interface Category {
  id: string;
  name: string;
  slug: string;
  description: string;
  sortOrder: number;
  postCount: number;
}

export type Group = PostGroup;

export interface Tag {
  id: string;
  name: string;
  slug: string;
  postCount: number;
}

export interface PostListParams {
  page?: number;
  pageSize?: number;
  groupId?: string;
  category?: string;
  tag?: string;
  keyword?: string;
  postType?: PostType;
  sort?: 'newest' | 'oldest';
}

export interface PostListData {
  list: Post[];
  total: number;
  page: number;
  pageSize: number;
}

export interface CreatePostData {
  title: string;
  postType?: PostType;
  visibility?: Visibility;
  templateKey?: string;
  templateData?: string;
  imageTextData?: ImageTextData;
  content: string;
  excerpt?: string;
  cover?: string;
  coverStorageKey?: string;
  groupId?: string;
  categoryId?: string;
  tagIds?: string[];
  status?: 'draft' | 'published' | 'archived';
  isTop?: boolean;
  publishNow?: boolean;
}

export interface BlogAIExcerptResponse {
  excerpt: string;
  model?: string;
}

export interface BlogAICoverResponse {
  prompt?: string;
  imageBase64?: string;
  imageUrl?: string;
  mimeType?: string;
  size?: string;
  model?: string;
}

export interface BlogReaderGuideResponse {
  guide: string;
  highlights: string[];
  path: string;
  model?: string;
}

export interface BlogAskCitation {
  heading: string;
  quote: string;
}

export interface BlogAskResponse {
  answer: string;
  citations?: BlogAskCitation[];
  model?: string;
}

export interface BlogRecommendItem {
  postId: string;
  title: string;
  excerpt: string;
  groupName?: string;
  readMinutes: number;
  reason: string;
}

export interface BlogRecommendResponse {
  items: BlogRecommendItem[];
  model?: string;
}

export function getPosts(params: PostListParams = {}) {
  return request.get<unknown, PostListData>('/public/blog/posts', { params });
}

export function getPostDetailById(id: string, config?: RequestConfig) {
  return request.get<unknown, PostDetail>(`/public/blog/posts/id/${id}`, config);
}

export function getPostComments(id: string, config?: RequestConfig) {
  return request.get<unknown, PostCommentListData>(`/public/blog/posts/id/${id}/comments`, config);
}

export function getAdminPostDetail(id: string, config?: RequestConfig) {
  return request.get<unknown, PostDetail>(`/admin/blog/posts/${id}`, config);
}

export function createPostComment(id: string, data: { content: string }) {
  return request.post<unknown, PostComment>(`/blog/posts/${id}/comments`, data);
}

export function deletePostComment(commentId: string) {
  return request.delete<unknown, { deleted: boolean }>(`/blog/comments/${commentId}`);
}

export function getCategories() {
  return request.get<unknown, Category[]>('/public/blog/categories');
}

export function getGroups(params: { authorId?: string; groupType?: GroupType } = {}) {
  return request.get<unknown, Group[]>('/public/blog/groups', { params });
}

export function getTags() {
  return request.get<unknown, Tag[]>('/public/blog/tags');
}

export function createPost(data: CreatePostData) {
  return request.post<unknown, Post>('/admin/blog/posts', data);
}

export function uploadImageTextAsset(formData: FormData) {
  return request.post<
    unknown,
    {
      url: string;
      key: string;
      fileName: string;
      size: number;
      width: number;
      height: number;
    }
  >('/admin/blog/image-text/assets/upload', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
}

export function uploadBlogCover(formData: FormData) {
  return request.post<
    unknown,
    {
      url: string;
      storageKey: string;
      fileName: string;
      size: number;
      width: number;
      height: number;
    }
  >('/admin/blog/cover/upload', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
}

export function uploadBlogCoverByUrl(data: { url: string }) {
  return request.post<
    unknown,
    {
      url: string;
      storageKey: string;
      fileName: string;
      size: number;
      width: number;
      height: number;
    }
  >('/admin/blog/cover/upload-by-url', data);
}

export function generateBlogExcerpt(data: { title?: string; content: string }) {
  return request.post<unknown, BlogAIExcerptResponse>('/admin/blog/ai/excerpt', data);
}

export function generateBlogCover(data: { title?: string; excerpt?: string; content: string }) {
  return request.post<unknown, BlogAICoverResponse>('/admin/blog/ai/cover', data);
}

export function generateBlogReaderGuide(postId: string) {
  return request.post<unknown, BlogReaderGuideResponse>(
    `/public/blog/posts/id/${postId}/ai/guide`,
    {},
  );
}

export function askBlogPost(postId: string, data: { question: string }) {
  return request.post<unknown, BlogAskResponse>(`/public/blog/posts/id/${postId}/ai/ask`, data);
}

export function recommendBlogPosts(data: {
  prompt: string;
  groupId?: string;
  keyword?: string;
  sort?: 'newest' | 'oldest';
}) {
  return request.post<unknown, BlogRecommendResponse>('/public/blog/ai/recommend', data);
}

export function updatePost(id: string, data: Partial<CreatePostData>) {
  return request.put<unknown, null>(`/admin/blog/posts/${id}`, data);
}

export function deletePost(id: string) {
  return request.delete<unknown, null>(`/admin/blog/posts/${id}`);
}

export function getAdminPosts(
  params: {
    page?: number;
    pageSize?: number;
    status?: string;
    postType?: PostType;
    groupId?: string;
  } = {},
) {
  return request.get<unknown, PostListData>('/admin/blog/posts', { params });
}

export function getAdminGroups(params: { groupType?: GroupType } = {}) {
  return request.get<unknown, Group[]>('/admin/blog/groups', { params });
}

export function createGroup(data: {
  name: string;
  groupType?: GroupType;
  slug?: string;
  description?: string;
  parentId?: string;
  sortOrder?: number;
}) {
  return request.post<unknown, Group>('/admin/blog/groups', data);
}

export function updateGroup(
  id: string,
  data: {
    name?: string;
    groupType?: GroupType;
    slug?: string;
    description?: string;
    parentId?: string;
    sortOrder?: number;
  },
) {
  return request.put<unknown, null>(`/admin/blog/groups/${id}`, data);
}

export function deleteGroup(id: string) {
  return request.delete<unknown, null>(`/admin/blog/groups/${id}`);
}
